processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -244,22 +243,57 @@ public int lookupId(String name)
     };
   }
 
+  public static class DimensionDesc
+  {
+    final int index;
+    final String name;
+    final DimDim values;
+    final ColumnCapabilitiesImpl capabilities;
+
+    public DimensionDesc(int index, String name, DimDim values, ColumnCapabilitiesImpl capabilities)
+    {
+      this.index = index;
+      this.name = name;
+      this.values = values;
+      this.capabilities = capabilities;
+    }
+  }
+
+  public static class MetricDesc
+  {
+    final int index;
+    final String name;
+    final String type;
+    final ColumnCapabilitiesImpl capabilities;
+
+    public MetricDesc(int index, AggregatorFactory factory)
+    {
+      this.index = index;
+      this.name = factory.getName();
+      this.type = factory.getTypeName();
+      this.capabilities = new ColumnCapabilitiesImpl();
+      if (type.equalsIgnoreCase("float")) {
+        capabilities.setType(ValueType.FLOAT);
+      } else if (type.equalsIgnoreCase("long")) {
+        capabilities.setType(ValueType.LONG);
+      } else {
+        capabilities.setType(ValueType.COMPLEX);
+      }
+    }
+  }
+
   private final long minTimestamp;
   private final QueryGranularity gran;
   private final List<Function<InputRow, InputRow>> rowTransformers;
-  private final AggregatorFactory[] metrics;
-  private final Map<String, Integer> metricIndexes;
-  private final Map<String, String> metricTypes;
-  private final ImmutableList<String> metricNames;
-  private final LinkedHashMap<String, Integer> dimensionOrder;
+  private final AggregatorFactory[] factories;
   private final AggregatorType[] aggs;
-  private final DimensionHolder dimValues;
-  private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
   private final boolean deserializeComplexMetrics;
 
-  protected final CopyOnWriteArrayList<String> dimensions;
+  private final Map<String, MetricDesc> metrics;
+  private final Map<String, DimensionDesc> dimensions;
+  private final Map<String, ColumnCapabilitiesImpl> columnCapabilities;
 
-  private volatile AtomicInteger numEntries = new AtomicInteger();
+  private transient final AtomicInteger numEntries = new AtomicInteger();
```

Comment:
I think removing "volatile" there should be okay, seems like AtomicInteger read/writes have the same memory effects as accesses on a volatile variable:

http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/atomic/package-summary.html